craft_application/secrets.py

```diff
@@ -0,0 +1,118 @@
+# This file is part of craft_application.
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Handling of build-time secrets."""
+from __future__ import annotations
+
+import re
+import subprocess
+from typing import Any
+
+from craft_application import errors
+
+SECRET_REGEX = re.compile(r"\$\(HOST_SECRET:(?P<command>.*)\)")
+
+
+def render_secrets(yaml_data: dict[str, Any]) -> set[str]:
+    """Render/expand the build secrets in a project's yaml data (in-place).
+
+    This function will process directives of the form $(HOST_SECRET:<cmd>) in string
+    values in ``yaml_data``. For each such directive, the <cmd> part is executed (with
+    bash) and the resulting output replaces the whole directive. The returned set
+    contains the outputs of all HOST_SECRET processing (for masking with craft-cli).
+
+    Note that only a few fields are currently supported:
+
+    - "source" and "build-environment" for parts.
```

Comment:
is there a specific reason for this?

Reply:
That's what we agreed on, back in the spec. I think we were "healthily afraid" of the potential consequences of enabling the feature everywhere.

... actually this is not completely true, we also agreed to enable this for `package-repositories`. However, when implementing this it occurred to me that we might want to restrict that scope even further: my suggestion is that when we know the overall "form" of Ubuntu Pro support in craft-archives we come back here and enable only the required field(s) (e.g. something like a `pro-key` field or similar).